Add function `core::iter::chain`

rossmacarthur opened the pull request:
The addition of `core::iter::zip` (#82917) set a precedent for adding plain functions for iterator adaptors. Adding `chain` makes it a little easier to `chain` two iterators.

```rust
for (x, y) in chain(xs, ys) {}
// vs.
for (x, y) in xs.into_iter().chain(ys) {}
```

There is prior art for the utility of this in [`itertools::chain`](https://docs.rs/itertools/latest/itertools/fn.chain.html).

Approved ACP https://github.com/rust-lang/libs-team/issues/154

rossmacarthur commented:
@rustbot label +T-libs-api -T-libs

Nilstrieb commented:
You need to create an [ACP](https://std-dev-guide.rust-lang.org/feature-lifecycle/api-change-proposals.html) first.

arniu commented on library/core/src/iter/adapters/chain.rs:
```diff
@@ -37,6 +37,39 @@ impl<A, B> Chain<A, B> {
+pub fn chain<A, B>(a: A, b: B) -> Chain<A::IntoIter, B::IntoIter>
+where
+    A: IntoIterator,
+    B: IntoIterator,
```
This line should be `B: IntoIterator<Item = A::Item>`

rossmacarthur replied:
🤦‍♂️

pitaj commented:
@rustbot label -S-waiting-on-author +S-waiting-on-ACP

joshtriplett commented:
r? libs

joshtriplett commented:
Correction:

r? libs-api

rossmacarthur commented:
@Amanieu I see the ACP was approved, let me know if this PR needs any changes.

rossmacarthur committed:
Add function `core::iter::chain`

The addition of `core::iter::zip` (#82917) set a precedent for adding
plain functions for iterator adaptors. Adding `chain` makes it a little
easier to `chain` two iterators.

```
for (x, y) in chain(xs, ys) {}
// vs.
for (x, y) in xs.into_iter().chain(ys) {}
```

rossmacarthur commented on library/core/src/iter/adapters/chain.rs:
````diff
@@ -37,6 +37,39 @@ impl<A, B> Chain<A, B> {
+/// assert_eq!(iter.next(), Some(6));
+/// assert_eq!(iter.next(), None);
+/// ```
+#[unstable(feature = "iter_chain", reason = "recently added", issue = "none")]
````
Done

Amanieu commented:
@bors r+

bors commented:
:pushpin: Commit 6a84995fae01eb0ae18ef6bfc41b2979ca760447 has been approved by `Amanieu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Amanieu 6a84995fae01eb0ae18ef6bfc41b2979ca760447 -->
<!-- homu: {"type":"Approved","sha":"6a84995fae01eb0ae18ef6bfc41b2979ca760447","approver":"Amanieu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
